source/ada/lsp-ada_handlers-refactor_change_parameter_mode.adb

```diff
@@ -0,0 +1,415 @@
+
+               if Key = "context" then
+                  LSP.Types.Read (JS, C.Context);
+
+               elsif Key = "where" then
+                  LSP.Messages.TextDocumentPositionParams'Read (JS, C.Where);
+
+               elsif Key = "first_parameter" then
+                  LSP.Types.Read (JS, C.First_Param_Index);
+
+               elsif Key = "last_parameter" then
+                  LSP.Types.Read (JS, C.Last_Param_Index);
+
+               elsif Key = "new_mode" then
+                  LSP.Types.Read (JS, C.New_Mode);
+
+               else
+                  JS.Skip_Value;
+               end if;
+            end;
+         end loop;
+
+         JS.R.Read_Next;
+      end return;
+   end Create;
+
+   -------------
+   -- Execute --
+   -------------
+
+   overriding
+   procedure Execute
+     (Self    : Command;
+      Handler : not null access LSP.Server_Notification_Receivers.
+        Server_Notification_Receiver'Class;
+      Client : not null access LSP.Client_Message_Receivers.
+        Client_Message_Receiver'Class;
+      Error : in out LSP.Errors.Optional_ResponseError)
+   is
+      Message_Handler : LSP.Ada_Handlers.Message_Handler renames
+        LSP.Ada_Handlers.Message_Handler (Handler.all);
+
+      Context         : LSP.Ada_Contexts.Context renames
+        Message_Handler.Contexts.Get (Self.Context).all;
+
+      Document : constant LSP.Ada_Documents.Document_Access :=
+        Message_Handler.Get_Open_Document (Self.Where.textDocument.uri);
+      Apply    : LSP.Messages.Client_Requests.Workspace_Apply_Edit_Request;
+
+      Workspace_Edits : LSP.Messages.WorkspaceEdit renames Apply.params.edit;
+
+      Node : constant Ada_Node :=
+        Document.Get_Node_At (Context, Self.Where.position);
+
+      Target_Subp               : Defining_Name := No_Defining_Name;
+      Target_Parameters_Indices : constant Parameter_Indices_Range_Type :=
+        (First => Positive (Self.First_Param_Index),
+         Last  => Positive (Self.Last_Param_Index));
+
+      Changer : Mode_Changer;
+      Edits   : Laltools.Refactor.Edit_Map;
+
+      function Analysis_Units return Analysis_Unit_Array is
+        (Context.Analysis_Units);
+      --  Provides the Context Analysis_Unit_Array to the Mode_Changer
+
+      function Get_Text_Document
+        (File_URI : LSP.Types.LSP_String)
+         return LSP.Messages.VersionedTextDocumentIdentifier;
+      --  Fetches a Document_Access given the File_URI and then creates a
+      --  OptionalVersionedTextDocumentIdentifier. This function is used as
+      --  an OptionalVersionedTextDocumentIdentifier provider to the
+      --  LSP.Lal_Utils.To_Workspace_Edit, which transforms an Edit_Map into
+      --  a WorkspaceEdit.
+
+      function Value (M : LSP.Types.LSP_String) return Ada_Mode;
+      --  Converts an LSP_String to an Ada_Mode
+
+      -----------------------
+      -- Get_Text_Document --
+      -----------------------
+
+      function Get_Text_Document
```

Comment:
To avoid code duplication we can have something like `Document_Version` method in the Message_Handler, I think.

Reply:
Agree, I'll add a `Get_Open_Document_Version` method.
`Message_Handler` implements `Document_Provider` interface.
Since `VersionedTextDocumentIdentifier` has a `Nullable_Number`, I'm adding this method to the `Document_Provider` interface too and `Document_Provider`s that do not have versioned documents will return a null number.